gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/DagManager.java

```diff
@@ -960,49 +962,76 @@ private void checkQuota(DagNode<JobExecutionPlan> dagNode) throws IOException {
+              proxyUser, specExecutorUri, getQuotaForUser(proxyUser),
+              proxyUserToJobCount.get(DagManagerUtils.getUserQuotaKey(proxyUser, dagNode))));
+        }
       }
 
       String serializedRequesters = DagManagerUtils.getSerializedRequesterList(dagNode);
       boolean requesterCheck = true;
-      String requesterMessage = null;
+
       if (serializedRequesters != null) {
         for (ServiceRequester requester : RequesterService.deserialize(serializedRequesters)) {
-          requesterCheck &= incrementMapAndCheckQuota(requesterToJobCount, requester.getName(), dagNode);
-          if (!requesterCheck && requesterMessage == null) {
-            requesterMessage = "Quota exceeded for requester " + requester.getName() + " on executor " + specExecutorUri + ": quota="
-                + getQuotaForUser(requester.getName()) + ", runningJobs=" + requesterToJobCount.get(DagManagerUtils.getUserQuotaKey(requester.getName(), dagNode));
+          int userQuotaIncrement = incrementJobCountAndCheckUserQuota(requesterToJobCount, requester.getName(), dagNode);
+          boolean thisRequesterCheck;
+          thisRequesterCheck = userQuotaIncrement < 0;  // user quota check failed
+          usersQuotaIncrement.put(requester.getName(), Math.max(userQuotaIncrement, 0));
+          requesterCheck = requesterCheck && thisRequesterCheck;
+          if (!thisRequesterCheck) {
+            requesterMessage.append(String.format(
+                "Quota exceeded for requester %s on executor %s : quota=%s, runningJobs=%d%n",
+                requester.getName(), specExecutorUri, getQuotaForUser(requester.getName()),
+                requesterToJobCount.get(DagManagerUtils.getUserQuotaKey(requester.getName(), dagNode))));
           }
         }
       }
 
       // Throw errors for reach quota at the end to avoid inconsistent job counts
-      if (!proxyUserCheck) {
-        throw new IOException("Quota exceeded for proxy user " + proxyUser + " on executor " + specExecutorUri +
-            ": quota=" + getQuotaForUser(proxyUser) + ", runningJobs=" + proxyUserToJobCount.get(DagManagerUtils.getUserQuotaKey(proxyUser, dagNode)));
-      }
+      if (!proxyUserCheck || !requesterCheck) {
+        if (!proxyUserCheck) {
+          decrementQuotaUsage(proxyUserToJobCount, proxyUser);
+        }
+
+        if (!requesterCheck) {
+          decrementQuotaUsageForUsers(usersQuotaIncrement);
+        }
 
-      if (!requesterCheck) {
-        throw new IOException(requesterMessage);
+        throw new IOException(requesterMessage.toString());
       }
     }
 
     /**
      * Increment quota by one for the given map and key.
-     * @return true if quota is not reached for this user or user is whitelisted, false otherwise.
+     * @return -1 if quota is already reached for this user, 0 if quota does not need to be increased, +1 if used quota is increased by 1
```

Comment:
as above, multiple threads may interleave increments and decrements, so if you'd like to log the actual number at the time the quota was exhausted, you'll have to return it here.  you could return the count as a negative number, with the sign acting as a sentinel of failure.

I'm comfortable w/ overloading the int return type w/ complex semantics, simply because Java lacks the Algebraic Data Types clearly suited to a direct/natural representation.  although, somewhat "clearer" to use (a full-fledged impl of) `Either` as we discussed earlier, this being a `private` method, I see no pressure to do so, and ultimately meager payoff.